code/modules/mob/living/carbon/human/human.dm

```diff
@@ -650,46 +650,65 @@
-						"<span class='notice'>You try to perform CPR on [C.name]... Hold still!</span>")
-		if(!do_mob(src, C))
-			to_chat(src, "<span class='warning'>You fail to perform CPR on [C]!</span>")
-			return 0
+	if (!getorganslot(ORGAN_SLOT_LUNGS))
+		to_chat(src, "<span class='warning'>You have no lungs to breathe with, so you cannot perform CPR!</span>")
+		return FALSE
 
-		var/they_breathe = !HAS_TRAIT(C, TRAIT_NOBREATH)
-		var/they_lung = C.getorganslot(ORGAN_SLOT_LUNGS)
+	if (HAS_TRAIT(src, TRAIT_NOBREATH))
+		to_chat(src, "<span class='warning'>You do not breathe, so you cannot perform CPR!</span>")
+		return FALSE
 
-		if(C.health > C.crit_threshold)
-			return
+	visible_message("<span class='notice'>[src] is trying to perform CPR on [target.name]!</span>", \
+					"<span class='notice'>You try to perform CPR on [target.name]... Hold still!</span>")
 
-		src.visible_message("<span class='notice'>[src] performs CPR on [C.name]!</span>", "<span class='notice'>You perform CPR on [C.name].</span>")
-		SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "perform_cpr", /datum/mood_event/perform_cpr)
-		C.cpr_time = world.time
-		log_combat(src, C, "CPRed")
-
-		if(they_breathe && they_lung)
-			var/suff = min(C.getOxyLoss(), 7)
-			C.adjustOxyLoss(-suff)
-			C.updatehealth()
-			to_chat(C, "<span class='unconscious'>You feel a breath of fresh air enter your lungs... It feels good...</span>")
-		else if(they_breathe && !they_lung)
-			to_chat(C, "<span class='unconscious'>You feel a breath of fresh air... but you don't feel any better...</span>")
-		else
-			to_chat(C, "<span class='unconscious'>You feel a breath of fresh air... which is a sensation you don't recognise...</span>")
+	if (!do_mob(src, target, time = panicking ? CPR_PANIC_SPEED : null))
+		to_chat(src, "<span class='warning'>You fail to perform CPR on [target]!</span>")
+		return FALSE
+
+	if (target.health > target.crit_threshold)
+		return FALSE
+
+	visible_message("<span class='notice'>[src] performs CPR on [target.name]!</span>", "<span class='notice'>You perform CPR on [target.name].</span>")
+	SEND_SIGNAL(src, COMSIG_ADD_MOOD_EVENT, "perform_cpr", /datum/mood_event/perform_cpr)
+	log_combat(src, target, "CPRed")
+
+	if (HAS_TRAIT(target, TRAIT_NOBREATH))
+		to_chat(target, "<span class='unconscious'>You feel a breath of fresh air... which is a sensation you don't recognise...</span>")
+	else if (!target.getorganslot(ORGAN_SLOT_LUNGS))
+		to_chat(target, "<span class='unconscious'>You feel a breath of fresh air... but you don't feel any better...</span>")
+	else
+		var/suff = min(target.getOxyLoss(), 7)
+		target.adjustOxyLoss(-suff)
+		target.updatehealth()
+		to_chat(target, "<span class='unconscious'>You feel a breath of fresh air enter your lungs... It feels good...</span>")
+
+	if (target.health <= target.crit_threshold)
+		if (!panicking)
+			to_chat(src, "<span class='warning'>[target] still isn't up! You try harder!</span>")
+		do_cpr(target, panicking = TRUE)
```

Comment:
Could you run a while instead of going recursive?

Reply:
Any particular reason for this? Does BYOND eventually stack overflow or something?

Reply:
Each proc call allocates memory. Potentially infinite recursion is a bad, bad idea.

There was this tale of BYOND only supporting a certain number of running procs at the same time, crashing when past those, but if that is still a thing or not I can't tell. Regardless, it's common good practice to make sure your possible recursion depth is finite (and within a sane value), and to prefer iterative solutions when it doesn't compromise clarity.

Reply:
In light of this I can rework it, yeah.